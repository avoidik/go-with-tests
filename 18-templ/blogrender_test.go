package blogrender_test

import (
	"blogrender"
	"bytes"
	"io"
	"testing"

	approvals "github.com/approvals/go-approval-tests"
)

func TestBlogRender(t *testing.T) {

	approvals.UseFolder("testdata")

	buf := bytes.Buffer{}
	aPost := blogrender.Post{
		Title:       "hello world",
		Body:        "This is **a post**",
		Description: "This is a description",
		Tags:        []string{"go", "tdd"},
	}

	rr, err := blogrender.NewPostRenderer()
	if err != nil {
		t.Fatal(err)
	}

	if err := rr.Render(&buf, aPost); err != nil {
		t.Fatal(err)
	}

	approvals.VerifyString(t, buf.String())
}

func TestRenderViewModel(t *testing.T) {
	approvals.UseFolder("testdata")

	buf := bytes.Buffer{}
	aPost := blogrender.Post{
		Title:       "hello world",
		Body:        "This is **a post**",
		Description: "This is a description",
		Tags:        []string{"go", "tdd"},
	}

	rr, err := blogrender.NewPostRenderer()
	if err != nil {
		t.Fatal(err)
	}

	if err := rr.RenderViewModel(&buf, aPost); err != nil {
		t.Fatal(err)
	}

	approvals.VerifyString(t, buf.String())
}

func TestIndexRender(t *testing.T) {

	approvals.UseFolder("testdata")

	buf := bytes.Buffer{}
	posts := []blogrender.Post{
		{Title: "First post"},
		{Title: "Second post"},
	}

	rr, err := blogrender.NewPostRenderer()
	if err != nil {
		t.Fatal(err)
	}

	if err := rr.RenderIndex(&buf, posts); err != nil {
		t.Error(err)
	}

	approvals.VerifyString(t, buf.String())
}

func BenchmarkBlogRender(b *testing.B) {
	aPost := blogrender.Post{
		Title:       "hello world",
		Body:        "This is a post",
		Description: "This is a description",
		Tags:        []string{"go", "tdd"},
	}

	rr, err := blogrender.NewPostRenderer()
	if err != nil {
		b.Fatal(err)
	}

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		rr.Render(io.Discard, aPost)
	}
}
