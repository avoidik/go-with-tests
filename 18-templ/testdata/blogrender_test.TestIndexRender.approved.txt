<!DOCTYPE html>
<html>
<head>
    <title>BLOG</title>
    <meta charset="UTF-8"/>
    <meta name="description" content="wonderful content" lang="en"/>
</head>
<body>
<nav role="navigation">
    <div>
        <h1>BLOG</h1>
        <ul>
            <li><a href="/">Home</a></li>
            <li><a href="/about">About</a></li>
            <li><a href="/archive">Archive</a></li>
        </ul>
    </div>
</nav>
<main>
<ol>
<li><a href="/post/first-post">First post</a></li>
<li><a href="/post/second-post">Second post</a></li>
</ol>
